inlang/source-code/sdk/src/persistence/slotfiles/createSlotWriter.test.ts

```diff
@@ -0,0 +1,145 @@
+import createSlotWriter from "./createSlotWriter.js"
+import { createNodeishMemoryFs } from "@lix-js/fs"
+
+type DocumentExample = {
+	id: string
+	content: string
+}
+
+// TODO write test if new file gets created on collison
+// TODO test that inserting the same item fails on memory
+// TODO test that inserting the same item (id) fails between processes
+// TODO test that data witten by another process reaches the process via fs events
+
+describe("Disconnected slot storage", () => {
+	it("insert should lead to a transient record", async () => {
+		const memoryFs = createNodeishMemoryFs()
+
+		const slotStorage = await createSlotWriter<DocumentExample>({
+			fs: memoryFs,
+			path: "/slot/",
+			fileNameCharacters: 3,
+			slotsPerFile: 16,
+			watch: false,
+		})
+		const insertedDocument: DocumentExample = {
+			id: "1",
+			content: "first document",
+		}
+		await slotStorage.insert(insertedDocument, false)
+
+		expect(slotStorage._writerInternals.transientSlotEntries.size).eq(1)
+		const docs1 = await slotStorage.findDocumentsById([insertedDocument.id])
+		expect(docs1.length).eq(1)
+		// expect(Object.isFrozen(docs1[0])).eq(true, "Records returned by slotStorage should be frozen")
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- we check the length before
+		expect(docs1[0]!.data.content).eq(insertedDocument.content)
+		const updateDocument = structuredClone(insertedDocument)
+		updateDocument.content = "updated"
+
+		await slotStorage.update(updateDocument, false)
+		expect(slotStorage._writerInternals.transientSlotEntries.size).eq(1)
+		const docs2 = await slotStorage.findDocumentsById([updateDocument.id])
+		expect(docs2.length).eq(1)
+		expect(docs2[0]!.data.content).eq(updateDocument.content)
+	})
+
+	it("update on a non existing record should throw", async () => {
+		const memoryFs = createNodeishMemoryFs()
+		const slotStorage = await createSlotWriter<DocumentExample>({
+			fs: memoryFs,
+			path: "/slot/",
+			fileNameCharacters: 3,
+			slotsPerFile: 16,
+			watch: false,
+		})
+		const nonExistingRecord: DocumentExample = {
+			id: "1",
+			content: "first document",
+		}
+		try {
+			await slotStorage.update(nonExistingRecord)
+		} catch (e) {
+			expect((e as Error).message).contain("does not exist")
+		}
+	})
+
+	it("insert should lead to a transient record save should lead to a record in non transient state", async () => {
+		const memoryFs = createNodeishMemoryFs()
+		const slotStorage = await createSlotWriter<DocumentExample>({
+			fs: memoryFs,
+			path: "/slot/",
+			fileNameCharacters: 3,
+			slotsPerFile: 16,
+			watch: false,
+		})
+
+		const insertedDocument: DocumentExample = {
+			id: "1",
+			content: "first document",
+		}
+
+		await slotStorage.insert(insertedDocument, false)
+
+		expect(slotStorage._writerInternals.transientSlotEntries.size).eq(1)
+
+		await slotStorage.save()
+		expect(slotStorage._writerInternals.transientSlotEntries.size).eq(0)
```

Comment:
aahhhhh so this is really a transient storage, not just a bad name for a memory layer on top of the file?! this is the first time i understand it is really working like that., maybe i misread this before?  if you just updated the memory representation of the file directly an then reconcile when saving you could reuse the same conflict logic as for writes by multiple threads and the whole impl would be simpler. i dont fully understand the need for this 3 layer design.